### PR TITLE
Improve text related to commit distribution

### PR DESCRIPTION
**Pull request by astrofrog**
Whoever wrote this text originally - please check that this is correct (is it # of commits vs authors or the other way around). I also rounded the value as suggested by @keflavich.

### CODE REVIEW

**Comment by adrn** [comment]
This would be @eteq - I'll ping him in slack

File: main.tex [hunk 1]
```diff
@@ -240,8 +240,9 @@ \subsection{Astropy development model}
 maintainability as the code itself.  Without comments there are $142197$ lines
 of code.} contributed by $232$ unique contributors over $19270$ \texttt{git}
 commits.
-The commit distribution \inlinecomment{Tom R}{too vague, maybe 'the distribution of number of commits as a function of ...'} follows an approximate power-law (with log-slope of
-$\approx -0.47$) \inlinecomment{AG}{round to -0.5?  This is astronomy, after all.}, meaning most of the commits were and are contributed by a
+The distribution of number of commits as a function of number of contributors
```
Comment (eteq) [inline main.tex]: Hmm... I would have said the original wording is correct. The data are not the number of commits as a function of <something>, but rather really just the probability distribution of the number of commits any given person has.

Although now that I think about it, I guess it's the same either way?  That is, I think the statement about "approximate power-law" and the value of the slope are the same whether you phrase it as a distribution or a binned function?

**Comment by eteq** [comment]
See my comment above on the wording... but agreed re: the rounding.

File: main.tex (hunk 1 above)
Comment (astrofrog) [reply main.tex]: The original wording was unclear, it just said 'The commit distribution', but it wasn't clear whether you meant the number of commits, and as a function of what. I guess what I mean is that 'The commit distribution' doesn't give me enough information about what x and y is :) Could you suggest an alternative re-wording?

Comment (eteq) [reply main.tex]: What about "The distribution of number of commits of each committer"?  Do you think that's clearer?  To me "the commit distribution" is a shorthand for exactly that.  That is to say, there is *no* "y" per se - it's a probability distribution over number of commits (although I guess you could consider "y" to be "probability"...).

Just to make sure we're on the same page, @astrofrog - the fact that it's not clear to you means it should be clarified - I'm just having trouble knowing what wording  to use because it *already* seems clear to me.

Comment (astrofrog) [reply main.tex]: @eteq - Just to clarify, I'm trying to understand what you are doing well enough to be able to reproduce the fit if I wanted.

Should I make a histogram where I bin contributors by number of commits? E.g. 10 authors made 1 commit, 3 authors made 2 commits, 1 author made 3 commits. Or am I showing number of commits versus commit rank? E.g. the most prolific commiter made 1000 commits, the second one made 100 commits, etc. What I'm trying to get at is that there are various diagnostic plots one could make with authors and committers. There has to be a y axis of some kind if you are fitting a power-law.

Just to be clear, I'm suggesting that 'the commit distribution' has as much information in it as 'the star distribution' or 'the galaxy distribution'. 'The commit distribution' could even mean over time.

"The distribution of number of commits of each committer" still doesn't really make it clear to me.

The wording here is especially important because we aren't showing the graph. If we were, then that would probably be immediately clearer. Could you show me the plot you mean?

Comment (astrofrog) [reply main.tex]: So you said 'it's a probability distribution over number of commits' which I think starts to address my question - but how did you do the fitting? Did you bin the commits?

Comment (astrofrog) [reply main.tex]: Anyway, show me the plot so I can understand why I'm not understanding :)

**Comment by astrofrog** [comment]
@eteq - what about now?

File: main.tex [hunk 2]
```diff
@@ -239,9 +239,11 @@ \subsection{Astropy development model}
 line count includes comments, as these are often as important for
 maintainability as the code itself.  Without comments there are $142197$ lines
 of code.} contributed by $232$ unique contributors over $19270$ \texttt{git}
-commits.
-The commit distribution \inlinecomment{Tom R}{too vague, maybe 'the distribution of number of commits as a function of ...'} follows an approximate power-law (with log-slope of
-$\approx -0.47$) \inlinecomment{AG}{round to -0.5?  This is astronomy, after all.}, meaning most of the commits were and are contributed by a
+commits. We can construct a distribution of the number of commits
```
Comment (adrn) [inline main.tex:242]: I still find this wording confusing. On re-reading, I think trying to describe this and stating a power-law index is more confusing than just showing the plot! So, my preference would be to (1) add the plot as a new figure, possibly 2-panel with 2nd panel showing cumulative number of committers over time, and (2) rephrase this to say something like:

> Figure\~X, left, shows the distribution of total number of commits per contributor as of XX 2017 in log-log space; from this, it's clear that most commits were and are contributed by a relatively small fraction of the contributor pool. Figure\~X, right, shows the number of contributors as a function of time since the genesis of the \astropypkg core package.

File: main.tex [hunk 3]
```diff
@@ -239,9 +239,11 @@ \subsection{Astropy development model}
...
+per contributor ordered by decreasing number of commits. When doing this,
+we find that this distribution
+follows an approximate power-law (with log-slope of
+$\approx -0.5$), meaning most of the commits were and are contributed by a
```
Comment (adrn) [inline main.tex:246]: Just to illustrate why this number is confusing and (I think) unnecessary: I'm wondering whether this -0.5 is the slope of the decreasing bins in `log(dN / dlog(Ncommits))` vs. `log(Ncommits)` space (is that what you mean by log-slope? because that sounds like `log(slope)`), or if it is the actual index of the power-law distribution? (which differs by +1)

**Comment by astrofrog** [comment]
@adrn - I agree with your comments. I would vote to either show the plot and clarify along the lines you said, or actually remove the mention of the power law and distribution and instead simply mention some basic stats as we do already (e.g. 50% of commits are from top 4 contributors etc)

**Review by eteq — APPROVED** [review]
LGTM now! (Had some out-of-band conversation with @astrofrog)

**Comment by eteq** [comment]
@adrn @astrofrog - I could imagine showing the plot if we think that's interesting... And indeed depending on how it's phrased exactly there's a +/- 1 difference. But I think it's important we show this, because there's no other place to talk about this, and when I show it to people they often think it's interesting/surprising.

**Comment by eteq** [comment]
(and I see @adrn's point here about the ambiguity. I mean the first thing you said, although in some literature that is *not* off-by-one from the distribution index... A cursory glance at some papers on this stuff in a social content suggested that to me, anyway, but of course I am no expert in that area...)

**Comment by adrn** [comment]
I agree it's important to mention the broader point, but don't think that hinges on the exact value of the power-law index! That is, if it were -0.942, I don't think people would be any more or less surprised 😄

**Comment by astrofrog** [comment]
@eteq - In the spirit of reproducible research I tried to remake the plot based on the current description and couldn't replicate the slope of -0.5:

commits

Here is my code:

```python
import subprocess
import numpy as np
import matplotlib.pyplot as plt

output = subprocess.check_output('git shortlog --summary --numbered', shell=True)
y = [int(line.split()[0]) for line in output.splitlines()]
x = np.arange(len(y)) + 1

plt.loglog(x, y)
plt.loglog(x, 1000 * x**-0.5)
plt.xlabel('Commiter rank')
plt.ylabel('Number of commits')
plt.savefig('commits.png')
```

Am I misunderstanding?

**Comment by adrn** [comment]
I could reproduce by instead just making a log-log histogram (`log(dN / dlog(Ncommits))`), but yea the description does not imply that this is where the number comes from:

image
(`n_commits` read from the shortlog)

```python
fig, ax = plt.subplots(1, 1, figsize=(6,5))

_bins = np.logspace(0, 3.5, 16)
ax.hist(n_commits, bins=_bins)
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('Number of commits')
ax.set_ylabel('Number of committers')

binctr = (_bins[:-1]+_bins[1:])/2.
ax.plot(binctr, 150*binctr**-0.5, marker='')
```

**Comment by astrofrog** [comment]
@adrn's plot is to me the most interesting one. Shall we update the text to refer to that?

**Comment by eteq** [comment]
yep, I agree adding @adrn's is the way to go.

Let this be a lesson to *me* to always do things in notebooks :wink:

**Comment by adrn** [comment]
See #29 - this can be closed if we go with that instead.

File: main.tex (hunk 2 above)
Comment (bsipocz) [reply main.tex:242]: +1 for the figures

**Comment by crawfordsm** [comment]
Closing this in favor of #29 -- Nice work here on the plots and analysis!
